add ova-compose

oliverkurth opened the pull request:
This adds a utility to create OVA files from simple configuration files in `yaml` format. This is more flexible than using static templates.

## Config File
```
system:
    name: example
    type: vmx-17
    os_vmw: other4xLinux64Guest
    firmware: efi
    secure_boot: true

networks:
    vm_network:
        name: "VM Network"
        description: "The VM Network network"

hardware:
    cpus: 2
    memory: 2048
    sata1:
        type: sata_controller
    scsi1:
        type: scsi_controller
    cdrom1:
        type: cd_drive
        parent: sata1
    rootdisk:
        type: hard_disk
        parent: scsi1
        disk_image: example.vmdk
    usb1:
        type: usb_controller
    ethernet1:
        type: ethernet
        subtype: VmxNet3
        network: vm_network
    videocard1:
        type: video_card
    vmci1:
        type: vmci

product:
    product: An Example VM
    vendor: A Company Inc.

annotation:
    text: the password is top secret

eula:
    file: eula.txt
```

The config file has 3 mandatory and 3 optional sections. `system`, `networks` and `hardware` are mandatory.
* `system` describes basic properties of the whole system, like name, hardware compatibility version and others.
* `networks` describes the network used. Each entry is a unique id. Each of these entries needs a `name` and a `description`.
* `hardware` describes the hardware components. Every entry is a unique id that can be any name, except the reserved ids `cpus` and `memory`. Each components except `cpus` and `memory` must have a `type`. The type can be one of the values described below.
  The reserved ids:
  * `cpus`: set to the number of CPUs
  * `memory`: set to the memory size in megabytes

  The other ids can have these types:
  *  `scsi_controller`, `sata_controller` or `ide_controller`: a controller. Each controller can have 0 or more other devices attached.
  * `scsi_controller` can have a `subtype` set to one of `VirtualSCSI` (aka "pvscsi") or `lsilogic`.
  * `cd_drive`: a CD drive, optionally with an ISO image set with `image`. The file will be packed within the OVA. The controller to attach to must be set with `parent` to the id of the controller. Set `connected = true` to have the image connected on startup (default is `false`)
  * `floppy`: a floppy device. Very similar to `cd_drive`, but does not need to be connected to a controller.
  * `hard_disk`: a hard disk. This must be set to an image in streamable vmdk format with `disk_image`. The file will be packed within the OVA.
  * `ethernet`: an ethernet device. The network must be set with `network` to one of the networks defined in the main `networks` section. Set `connected = false` to have the device disconnected on startup (default is `true`)
  * `usb_controller`, `video_card` and `vmci`: USB controller, video card and VMCI device. 

These 3 sections are optional:
* `product` describes the product. It has the fields `info`, `product`, `vendor`, `version` and `full_version`.
* `annotation` has the fields `info`, `text` and `file`. `text` and `file` are mutually exclusive - `text` is text inline, `file` can be set to a text file that will be filled in. The annotation text will appear for example as a comment in VMware Fusion.
* `eula` also has the fields `info`, `text` and `file`. It contains the EULA agreement the user has to agree to when deploying the VM.

## Parameters

Values can be filled in with parameters from the command line. This makes the yaml file reusable for different VMs. For example, the hard disk can be set via command line:
```
    rootdisk:
        type: hard_disk
        parent: scsi1
        disk_image: !param rootdisk
```
When invoking `ova-compose` (see below) the value can be set with `--param rootdisk=example.vmdk`.

Default values can also be set in case `ova-compose` is invoked without the parameter. Example:
```
hardware:
    cpus: !param cpus=2
    memory: !param memory=2048
```
In this case, setting the parameters `cpus` and `memory` will be optional.

If `ova-compose` is invoked without setting a parameter for which no default is set, it will throw an error.

## Usage

`ova-compose -i|--input_file <input_file> -o|--output_file <output_file> [ --format <format> ] [[--param <key=value>] ...] [-q]`
Options:
* `-i|--input_file <input_file>` : the config file to use
* `-o|--output_file <output_file>`: the output file or directory
* `--format <format>` : the format, one of: `ova`, `ovf` or `dir`. If not set, the format will be guessed from the output file extension if it is `ova` or `ovf`
  * `ova` to create an OVA file
  * `ovf` to create just the OVF file
  * `dir` to create a directory containing the OVF file, the manifest and the files used for the cdrom and harddisk devices.
* `--param <key=value>`: set parameter `<key>` to `<value>`.
 
Example:
```
$ ova-compose.py -i minimal.yaml -o minimal.ova
creating 'minimal.ova' with format 'ova' from 'minimal.yaml'
done.
```

keirazhang commented on ovf/ova-compose.py:
```diff
@@ -0,0 +1,746 @@
+
+class RasdIdeController(RasdController):
+    resource_type = 4
+    description = "IDE Controller"
```
After resolving network section, it reported another error:
```
Issues detected with selected template. Details: - 40:7:ELEMENT_ILLEGAL: Multiple ''memory'' elements are not allowed. - 77:7:VALUE_ILLEGAL: Value ''5'' of Parent element does not refer to a ref of type DiskControllerReference.
```

```
      <Item>
        <rasd:Description>IDE Controller</rasd:Description>
        <rasd:ElementName>ide1</rasd:ElementName>
        <rasd:InstanceID>5</rasd:InstanceID>
        <rasd:ResourceType>4</rasd:ResourceType>
      </Item>
```
By comparing ovf template, the resource type of IDE controller shall be 5.

keirazhang reviewed (commented):
I suggest adding the example config file as a sample for using ova-compose, and change default settings with below recommendations:

1. Change the hardware version to `vmx-17` as vmx-17 is the default hardware version of ESXi 7.0, which is the oldest version in general support at present.
2. Change the OS type to `other4xLinux64Guest`.
3. Change the firmware to `efi`.
4. Add secureboot config and set default value to false.
5. Change disk1 controller to SCSI.
6. Change cdrom1 controller to SATA.

oliverkurth commented on ovf/ova-compose.py:
```diff
@@ -0,0 +1,746 @@
+        self.hardware_config = OVF.CONFIG_DEFAULTS.copy()
+        self.name = system['name']
+        if 'firmware' in system:
+            self.hardware_config['firmware'] = system['firmware']
```
Done.

oliverkurth commented:
Thank you Qi for your comments. I think I addressed them all.

I made some additional changes:
* `system.os` is now `system.os_vmw` - I ewant to account for the fact that the ids are VMware specific. There is also `system.os_cim` now that can be set to the "CIM_OperatingSystem.OsType" . The default is `100` for Linux.
* I moved the script into its own directory
* it will be installed into `/usr/bin` (or whatever the prefix is set to)
